crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -177,45 +83,67 @@ pub struct CheckpointExecutorEventLoop {
     /// the final checkpoint will not have been set, thus CheckpointExecutor
     /// will reschedule the last checkpoint and correctly set end_of_epoch.
     end_of_epoch: bool,
```

Comment:
Why do we still need end_of_epoch? Shouldn't we reject any checkpoint received that's for a different epoch?

Reply:
`end_of_epoch` is for executor to know that we have already received a last checkpoint of epoch from state sync, so that it will not schedule any checkpoints past that for execution. In essence, it means "join on all existing checkpoint execution tasks". So we still need it

Reply:
hmm I haven't looked closely but I was thinking if it's easier to reason about if we simply don't schedule anything that's from a different epoch than the current epoch. And the exit condition is when the highest executed checkpoint equals to the last checkpoint of the epoch

Reply:
I see. Yeah that makes sense if we're passing in the epoch from `SuiNode`. Sounds good